src/librustc_incremental/calculate_svh/svh_visitor.rs

```diff
@@ -13,43 +13,172 @@
+        let (line_start, line_end) = file.line_bounds(line_index);
+
+        // Just overwrite some cache entry. If we got this for, all of them
+        // point to the wrong file.
+        self.line_cache[self.eviction_index] = (line_index + 1,
+                                                line_start,
+                                                line_end,
+                                                file.clone());
+        self.eviction_index = (self.eviction_index + 1) % self.line_cache.len();
+
+        return (file, line_index + 1, pos - line_start);
+    }
 }
 
 impl<'a, 'hash, 'tcx> StrictVersionHashVisitor<'a, 'hash, 'tcx> {
     pub fn new(st: &'a mut SipHasher,
                tcx: TyCtxt<'hash, 'tcx, 'tcx>,
-               def_path_hashes: &'a mut DefPathHashes<'hash, 'tcx>)
+               def_path_hashes: &'a mut DefPathHashes<'hash, 'tcx>,
+               hash_spans: bool)
                -> Self {
-        StrictVersionHashVisitor { st: st, tcx: tcx, def_path_hashes: def_path_hashes }
+        StrictVersionHashVisitor {
+            st: st,
+            tcx: tcx,
+            def_path_hashes: def_path_hashes,
+            hash_spans: hash_spans,
+            codemap: CachedCodemapView::new(tcx),
+        }
     }
 
     fn compute_def_id_hash(&mut self, def_id: DefId) -> u64 {
         self.def_path_hashes.hash(def_id)
     }
+
+    // Hash a span in a stable way. If we would just hash the spans BytePos
+    // fields that would be similar hashing pointers since those or just offsets
+    // into the CodeMap. Instead, we hash the (file name, line, column) triple,
+    // which stays the same even if the containing FileMap has moved within the
+    // CodeMap.
+    // Also note that we are hashing byte offsets for the column, not unicode
+    // codepoint offsets. For the purpose of the hash that's sufficient.
+    fn hash_span(&mut self, span: Span) {
+        debug_assert!(self.hash_spans);
+        debug!("hash_span: st={:?}", self.st);
+
+        // If this is not an empty or invalid span, we want to hash the last
+        // position that belongs to it, as opposed to hashing the first
+        // position past it.
+        let span_hi = if span.hi > span.lo {
+            // We might end up in the middle of a multibyte character here,
+            // but that's OK, since we are not trying to decode anything at
+            // this position.
+            span.hi - BytePos(1)
+        } else {
+            span.hi
+        };
+
+        let (file1, line1, col1) = self.codemap.byte_pos_to_line_and_col(span.lo);
+        let (file2, line2, col2) = self.codemap.byte_pos_to_line_and_col(span_hi);
+
+        let expansion_kind = match span.expn_id {
+            NO_EXPANSION => SawSpanExpnKind::NoExpansion,
+            COMMAND_LINE_EXPN => SawSpanExpnKind::CommandLine,
+            _ => SawSpanExpnKind::SomeExpansion,
+        };
+
+        expansion_kind.hash(self.st);
```

Comment:
TODO: Remove this line. The expansion kind is already hashed later as field of `SawSpan`.